v2/apiserver/internal/core/kubernetes/substrate.go

```diff
@@ -251,8 +278,147 @@ func (s *substrate) DeleteProject(
 	return nil
 }
 
-func generateNewNamespace(projectID string) string {
-	return fmt.Sprintf("brigade-%s-%s", projectID, uuid.NewV4().String())
+func (s *substrate) ScheduleWorker(
+	ctx context.Context,
+	project core.Project,
+	event core.Event,
+) error {
+	// Create a Kubernetes secret containing relevant Event and Project details.
+	// This is created PRIOR to scheduling so that these details will reflect an
+	// accurate snapshot of Project configuration at the time the Event was
+	// created.
+
+	projectSecretsSecret, err := s.kubeClient.CoreV1().Secrets(
+		project.Kubernetes.Namespace,
+	).Get(ctx, "project-secrets", metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(
+			err,
+			"error finding secret \"project-secrets\" in namespace %q",
+			project.Kubernetes.Namespace,
+		)
+	}
+	secrets := map[string]string{}
+	for key, value := range projectSecretsSecret.Data {
+		secrets[key] = string(value)
+	}
+
+	type proj struct {
+		ID         string                 `json:"id"`
+		Kubernetes core.KubernetesDetails `json:"kubernetes"`
+		Secrets    map[string]string      `json:"secrets"`
+	}
+
+	type worker struct {
+		APIAddress           string            `json:"apiAddress"`
+		APIToken             string            `json:"apiToken"`
+		LogLevel             core.LogLevel     `json:"logLevel"`
+		ConfigFilesDirectory string            `json:"configFilesDirectory"`
+		DefaultConfigFiles   map[string]string `json:"defaultConfigFiles" bson:"defaultConfigFiles"` // nolint: lll
+	}
+
+	// Create a secret with event details
+	eventJSON, err := json.MarshalIndent(
+		struct {
```

Comment:
Q: why not use the `Event` struct here?

Reply:
Good question. Initially I did, but the Event struct only had a reference to a project (projectID) and I want the worker to have access to (some) project details. Options are use a different struct that encapsualtes a copy of the (very few) relevant project details or else use a completely separate chunk of JSON, which seemed it would complicate things a bit on the brigade.js end.

Reply:
@vdice Amending my previous answer... I was thinking about this more and I think I really need to justify why replicating some of the `Event` fields in a second struct does not constitute a maintenance burden going forward. i.e. Can I convince myself that every time a field gets (theoretically) added to the `Event` struct that it doesn't automatically _have_ to be replicated here as well.

I believe what we have here is a _worker-specific representation_ of an `Event`. Some `Event` fields are not of interest to a worker. A worker also needs to know certain _extra_ things that aren't part of the `Event` struct. The example I gave it the previous comment included project details (including secrets), but there are others as well. The `Worker` field here, for instance, has a field of its own-- `APIAddress`-- which isn't part of an `Event` but is information that a worker requires in order to _process_ the event...

Worker-specific event representation.

Reply:
Definitely.  Thanks for the explanation.